src/adjustproxyconnection.cpp

```diff
@@ -0,0 +1,160 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "adjustproxyconnection.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "qmlengineholder.h"
+#include "networkrequest.h"
+
+#include <QFile>
+#include <QHostAddress>
+#include <QTcpSocket>
+
+constexpr const char* HTTP_200_RESPONSE =
+    "HTTP/1.1 200 OK\nContent-Type: "
+    "application/json\n\n{}\n";
+
+namespace {
+Logger logger(LOG_MAIN, "AdjustProxyConnection");
+}  // namespace
+
+AdjustProxyConnection::AdjustProxyConnection(QObject* parent,
+                                             QTcpSocket* connection)
+    : QObject(parent), m_connection(connection) {
+  MVPN_COUNT_CTOR(AdjustProxyConnection);
+
+  logger.debug() << "New connection received";
+
+  Q_ASSERT(m_connection);
+  connect(m_connection, &QTcpSocket::readyRead, this,
+          &AdjustProxyConnection::readData);
+}
+
+AdjustProxyConnection::~AdjustProxyConnection() {
+  MVPN_COUNT_DTOR(AdjustProxyConnection);
+  logger.debug() << "Connection released";
+}
+
+void AdjustProxyConnection::readData() {
+  Q_ASSERT(m_connection);
+  QByteArray input = m_connection->readAll();
+  m_buffer.append(input);
+
+  // Parsing method and route
+  int pos = m_buffer.indexOf("\n");
+  if (pos == -1) {
+    return;
+  }
+
+  QByteArray line = m_buffer.left(pos);
+  m_buffer.remove(0, pos + 1);
+
+  QList<QByteArray> parts = line.split(' ');
+  if (parts.length() < 2) {
+    return;
+  }
+
+  m_method.append(parts[0]);
+  m_route.append(parts[1]);
+
+  if (m_method == "GET") {
+    pos = m_route.indexOf('?');
+    if (pos > -1) {
+      m_parametersString = m_route.right(m_route.length() - (pos + 1));
+      m_route.remove(pos, m_route.length());
+    }
+  }
+
+  // Parsing headers
+  while (true) {
+    pos = m_buffer.indexOf("\n");
+    if (pos == -1) {
+      break;
+    }
+
+    QByteArray line = m_buffer.left(pos);
+    m_buffer.remove(0, pos + 1);
+
+    QByteArray header = line.trimmed();
+    if (header == "") {
+      break;
+    }
+
+    pos = header.indexOf(":");
+    if (pos == -1) {
+      continue;
+    }
+
+    QByteArray headerName = header.left(pos);
+    header.remove(0, pos + 1);
+    QByteArray headerValue = header.trimmed();
+
+    m_headers.insert(headerName, headerValue);
```

Comment:
We do need to parse the headers just to retrieve the Content-Length one. All the rest should be kept and stored as it is.
The Content-Length header is important because we have to read that amount of data from the buffer after the end of the header "\n\n" or "\r\n\r\n".